python/ray/autoscaler/_private/resource_demand_scheduler.py

```diff
@@ -115,9 +126,59 @@ def get_nodes_to_launch(
         total_nodes_to_add = self._get_concurrent_resource_demand_to_launch(
             total_nodes_to_add, usage_by_ip.keys(), nodes, pending_nodes)
 
+        if self.is_legacy_yaml:
+            assert len(total_nodes_to_add.keys()) <= 1
         logger.info("Node requests: {}".format(total_nodes_to_add))
         return total_nodes_to_add
 
+    def _handle_legacy_yaml(self, nodes: List[NodeID],
+                            pending_nodes: Dict[NodeType, int],
+                            static_node_resources: List[ResourceDict]
+                            ) -> (bool, Dict[NodeType, int]):
+        """Handles legacy config files without available_node_types.
+
+        Launches a single node from which it calculates the resources and fills
+        the available self.node_types.
+        Args:
+            nodes: List of existing nodes in the cluster.
+            pending_nodes: Summary of node types currently being launched.
+            static_node_resources: List of static node resources.
+        Returns:
+            legacy_nodes_to_add: A dict that is either empty or includes
+                max(1, min_workers) nodes to add from which we later calculate
+                the resources.
+            return_immediately: Whether to return legacy_nodes_to_add or to
+                continue running the resource demand scheduler.
+        """
+        return_immediately, legacy_nodes_to_add = False, {}
+        if not self.node_types:
+            if self.cluster_config["max_workers"] == 0:
+                # Do not add any workers if max_workers is 0.
+                return_immediately, legacy_nodes_to_add = True, {}
```

Comment:
Shouldn't we also return zero if there isn't any queue? Otherwise we will launch a new node for no reason even if the head node itself can handle the workload.

Reply:
Yes.

Reply:
I think this can be implemented by simply checking if `resource_demands` is empty.